common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -533,14 +533,19 @@ public boolean contentEqualsIgnoreCase(CharSequence string) {
             return false;
         }
 
+        byte[] value = this.value;
         if (string instanceof AsciiString) {
             AsciiString rhs = (AsciiString) string;
-            for (int i = arrayOffset(), j = rhs.arrayOffset(), end = i + length(); i < end; ++i, ++j) {
-                if (!equalsIgnoreCase(value[i], rhs.value[j])) {
-                    return false;
+            byte[] rhsValue = rhs.value;
+            if ((arrayOffset() | rhs.arrayOffset()) == 0) {
+                for (int i = 0; i < length; ++i) {
+                    if (!equalsIgnoreCase(value[i], rhsValue[i])) {
```

Comment:
Let me extract it into a thread:

> And I think it could be made even faster by using AsciiString::cache to perform a quick isLowerCase on the cached AsciiString saving a byte to know if is a full lowercase String aready (which is the most common use case for cached ones).
This will enable super quick equality checks for known HTTP 1.1 lower-case header names and faster ones for cases like the one used in the benchmark (eg Accept vs accept). wdyt @normanmaurer @idelpivnitskiy ?


This sounds like an interesting idea, but possible for another PR, bcz adding a knowledge of "all lower case" or "all upper case" may help to improve a few other methods.

I also wonder if we need to implement a real internal cache (with `ConcurrentHashMap`?) for `AsciiString::cached` method to always return the same `AsciiString` instance when users try to create an `AsciiString` equivalent for exact same `String`. wdyt?

Reply:
> This sounds like an interesting idea, but possible for another PR, bcz adding a knowledge of "all lower case" or "all upper case" may help to improve a few other methods.

yep, I think it would be super worthy, given that we currently perform this check on HTTP 2 as well and if the `AsciiString` would be known to be already all lower-case, will skip it.

> I also wonder if we need to implement a real internal cache (with ConcurrentHashMap?) for AsciiString::cached method to always return the same AsciiString instance when users try to create an AsciiString equivalent for exact same String. wdyt?

I have no idea about this one, maybe @normanmaurer got some historical data points why it hasn't been like that already (although CHM has been introduced later TBH).

Reply:
I have no idea either :D